core/src/Streamly/Internal/Data/Serialize.hs

```diff
@@ -7,272 +12,107 @@
 
-    -- {-# INLINE size #-}
-    size acc xs = foldl' size (acc + (Unbox.sizeOf (Proxy :: Proxy Int))) xs
+unliftInteger :: LiftedInteger -> Integer
+unliftInteger (LIS (I# x)) = IS x
+unliftInteger (LIP (Array (MutableByteArray x) _ _)) = IP (unsafeCoerce# x)
+unliftInteger (LIN (Array (MutableByteArray x) _ _)) = IN (unsafeCoerce# x)
 
-    -- Inlining this causes large compilation times for tests
-    {-# INLINABLE deserialize #-}
-    deserialize off arr sz = do
-        (off1, len) <- deserialize off arr sz :: IO (Int, Int)
-        let
-            peekList f o i | i >= 3 = do
-              -- Unfold the loop three times
-              (o1, x1) <- deserialize o arr sz
-              (o2, x2) <- deserialize o1 arr sz
-              (o3, x3) <- deserialize o2 arr sz
-              peekList (f . (\xs -> x1:x2:x3:xs)) o3 (i - 3)
-            peekList f o 0 = pure (o, f [])
-            peekList f o i = do
-              (o1, x) <- deserialize o arr sz
-              peekList (f . (x:)) o1 (i - 1)
-        peekList id off1 len
+#else
 
-    -- Inlining this causes large compilation times for tests
-    {-# INLINABLE serialize #-}
-    serialize off arr val = do
-        void $ serialize off arr (length val)
-        let off1 = off + Unbox.sizeOf (Proxy :: Proxy Int)
-        let pokeList o [] = pure o
-            pokeList o (x:xs) = do
-              o1 <- serialize o arr x
-              pokeList o1 xs
-        pokeList off1 val
+liftInteger :: Integer -> LiftedInteger
+liftInteger (S# x) = LIS (I# x)
+liftInteger (Jp# (BN# x)) =
+    LIP (Array (MutableByteArray (unsafeCoerce# x)) 0 (I# (sizeofByteArray# x)))
+liftInteger (Jn# (BN# x)) =
+    LIN (Array (MutableByteArray (unsafeCoerce# x)) 0 (I# (sizeofByteArray# x)))
 
---------------------------------------------------------------------------------
--- High level functions
---------------------------------------------------------------------------------
+unliftInteger :: LiftedInteger -> Integer
+unliftInteger (LIS (I# x)) = S# x
+unliftInteger (LIP (Array (MutableByteArray x) _ _)) =
+    Jp# (BN# (unsafeCoerce# x))
+unliftInteger (LIN (Array (MutableByteArray x) _ _)) =
+    Jn# (BN# (unsafeCoerce# x))
 
-{-# INLINE encodeAs #-}
-encodeAs :: forall a. Serialize a => PinnedState -> a -> Array Word8
-encodeAs ps a =
-    unsafeInlineIO $ do
-        let len = size 0 a
-        mbarr <- Unbox.newBytesAs ps len
-        off <- serialize 0 mbarr a
-        assertM(len == off)
-        pure $ Array mbarr 0 off
+#endif
 
-{-# INLINE encode #-}
-encode :: Serialize a => a -> Array Word8
-encode = encodeAs Unpinned
+instance Serialize Integer where
+    {-# INLINE size #-}
+    size i a = size i (liftInteger a)
 
-{-# INLINE pinnedEncode #-}
-pinnedEncode :: Serialize a => a -> Array Word8
-pinnedEncode = encodeAs Pinned
+    {-# INLINE deserialize #-}
+    deserialize off arr end = fmap unliftInteger <$> deserialize off arr end
```

Comment:
I am wondering if it will be simpler to implement it directly rather than using lift/unlift.

Reply:
I'll try implementing it directly. This is just an indirection for convinence.

Reply:
See : https://github.com/composewell/streamly/issues/2582